src/UseCase/BannerSelection/BannerSelectionUseCase.php

```diff
@@ -4,29 +4,48 @@
 
 namespace WMDE\BannerServer\UseCase\BannerSelection;
 
+use WMDE\BannerServer\Entity\BannerSelection\Bucket;
+use WMDE\BannerServer\Entity\BannerSelection\Campaign;
 use WMDE\BannerServer\Entity\BannerSelection\CampaignCollection;
+use WMDE\BannerServer\Entity\BannerSelection\ImpressionThreshold;
+use WMDE\BannerServer\Utils\RandomIntegerInterface;
 
 /**
  * @license GNU GPL v2+
  */
 class BannerSelectionUseCase {
 
 	private $campaignCollection;
+	private $currentCampaign;
+	private $impressionThreshold;
 
-	public function __construct( CampaignCollection $campaignCollection ) {
+	public function __construct( CampaignCollection $campaignCollection, ImpressionThreshold $impressionThreshold ) {
 		$this->campaignCollection = $campaignCollection;
+		$this->impressionThreshold = $impressionThreshold;
 	}
 
 	public function provideBannerRequest( Visitor $visitor ): BannerSelection {
-		if ( $visitor->hasDonated() ) {
+		if ( $visitor->hasDonated() ||
+			$this->getCurrentCampaign() === null ||
+			$this->impressionThreshold->isThresholdReached( $visitor->getTotalImpressionCount() ) ) {
```

Comment:
Please add a check for `$this->getCurrentCampaign()->isRatioLimited()` 

you might also select `$this->currentCampaign` in the constructor and drop the `campaignCollection` property. Then you could even get rid of the `getCurrentCampaign()` method.